gnucash/gnome-utils/gnc-main-window.cpp

```diff
@@ -3572,60 +3685,142 @@ gnc_main_window_unmerge_actions (GncMainWindow *window,
     if (entry == nullptr)
         return;
 
-    gtk_ui_manager_remove_action_group (window->ui_merge, entry->action_group);
-    gtk_ui_manager_remove_ui (window->ui_merge, entry->merge_id);
-    gtk_ui_manager_ensure_update (window->ui_merge);
+    gtk_widget_insert_action_group (GTK_WIDGET(window), group_name, nullptr);
 
     g_hash_table_remove (priv->merged_actions_table, group_name);
 }
 
 
-/*  Force a full update of the user interface for the specified
- *  window.  This can be an expensive function, but is needed because
- *  the gtk ui manager doesn't always seem to update properly when
- *  actions are changed.
- */
-void
-gnc_main_window_actions_updated (GncMainWindow *window)
+struct group_iterate
 {
-    GtkActionGroup *force;
+    GAction *action;
+    const gchar *group_name;
+    const gchar *action_name;
+};
 
-    g_return_if_fail (GNC_IS_MAIN_WINDOW (window));
+static void
+group_foreach_cb (gpointer key, gpointer item, gpointer arg)
+{
+    struct group_iterate *iter = static_cast<group_iterate*>(arg);
+    GAction *action;
 
-    /* Unfortunately gtk_ui_manager_ensure_update doesn't work
-     * here.  Force a full update by adding and removing an empty
-     * action group.
-     */
-    force = gtk_action_group_new("force_update");
-    gtk_ui_manager_insert_action_group (window->ui_merge, force, 0);
-    gtk_ui_manager_ensure_update (window->ui_merge);
-    gtk_ui_manager_remove_action_group (window->ui_merge, force);
-    g_object_unref(force);
-}
+    if (iter->action)
+        return;
+
+    if (iter->group_name != nullptr)
+    {
+        if (g_strcmp0 (iter->group_name, (gchar*)key) != 0)
+            return;
+    }
+
+    auto entry{static_cast<MergedActionEntry *>(item)};
+
+    action = g_action_map_lookup_action (G_ACTION_MAP(entry->simple_action_group),
+                                         iter->action_name);
 
+    if (action)
+        iter->action = action;
+}
 
-GtkAction *
+GAction *
 gnc_main_window_find_action (GncMainWindow *window, const gchar *name)
 {
-    GtkAction *action = nullptr;
-    const GList *groups, *tmp;
+    GncMainWindowPrivate *priv;
+    GAction *action = nullptr;
+
+    priv = GNC_MAIN_WINDOW_GET_PRIVATE(window);
+
+    action = g_action_map_lookup_action (G_ACTION_MAP(priv->simple_action_group),
+                                         name);
+
+    return action;
+}
+
+GAction *
+gnc_main_window_find_action_in_group (GncMainWindow *window,
+                                      const gchar *group_name,
+                                      const gchar *name)
+{
+    GncMainWindowPrivate *priv;
+    GAction *action = nullptr;
+    struct group_iterate iter;
+
+    priv = GNC_MAIN_WINDOW_GET_PRIVATE(window);
+
+    if (priv->merged_actions_table == nullptr)
+        return nullptr;
+
+    iter.action = nullptr;
+    iter.action_name = name;
+    iter.group_name = group_name;
+
+    g_hash_table_foreach (priv->merged_actions_table, group_foreach_cb, &iter);
```

Comment:
I'm not convinced we need to keep track of merged actions in a merged_actions_table. So I'm systematically going through all uses and check if there are alternatives.
I *think* in this case what you really want to do is this:
```
auto action_group = gtk_widget_get_action_group (GTK_WIDGET(window), group_name);
if (action_group)
    action = g_action_map_lookup_action (G_ACTION_MAP(action_group), name);
```